binary_search_tree/tree.py

```diff
@@ -9,40 +9,121 @@ def __init__(self, key, val = None):
         self.right = None
         
 
-
 class Tree:
     def __init__(self):
         self.root = None
 
+    def addHelper(self, current_root, new_node):
+        if current_root is None:
+            return new_node
+        
+        if new_node.key < current_root.key:
+            if current_root.left is None:
+                # current_root.left = self.addHelper(current_root.left, new_node)
+                current_root.left = new_node
+            else:
+                self.addHelper(current_root.left, new_node)
+        else:
+            if current_root.right is None:
+                # current_root.right = self.addHelper(current_root.right, new_node)
+                current_root.right = new_node
+            else:
+                self.addHelper(current_root.right, new_node)
+
     # Time Complexity: 
     # Space Complexity: 
     def add(self, key, value = None):
-        pass
+        new_node = TreeNode(key, value)
+
+        if self.root is None:
+            self.root = new_node
+            return 
+        self.addHelper(self.root, new_node)
+
+    
+    def findHelper(self, current_node, key):
+        if not current_node:
+            print('in first case')
+            return None
+        if current_node.key == key:
+            print('in found case')
+            return current_node.value
+        if current_node.key > key:
+            print('in left case')
+            return self.findHelper(current_node.left, key)
+        if current_node.key < key:
+            print('in right case')
+            return self.findHelper(current_node.right, key)    
 
     # Time Complexity: 
-    # Space Complexity: 
+    # Space Complexity:
     def find(self, key):
-        pass
+        #if it's empty return None
+        return self.findHelper(self.root, key)
+
+    def inorderHelper(self, current, result):
+        if current:
+            self.inorderHelper(current.left, result)
+            result.append({"key": current.key, "value": current.value})
+            self.inorderHelper(current.right, result)
 
     # Time Complexity: 
     # Space Complexity: 
     def inorder(self):
-        pass
+        #only instantiate once
+        result = []
+        self.inorderHelper(self.root, result)
+        return result
+
 
     # Time Complexity: 
-    # Space Complexity:     
+    # Space Complexity: 
+    def preorderHelper(self, current, result):
+        #The "shape" -> "<"
+        if current:
+            result.append({"key": current.key, "value": current.value})
+            self.preorderHelper(current.left, result)
+            self.preorderHelper(current.right, result)
+
+
     def preorder(self):
```

Comment:
👀 Good job implementing the function. But what would the time and space complexity be? 🤔